docs/design/persistent-storage.md

```diff
@@ -0,0 +1,235 @@
+
+`PersistenceRequirements` expresses the capabilities and performance attributes of the underlying VolumeSource.  Allow discovery
+of storage by its capabilities and use those capabilities as pod storage requirements.
+
+Kubernetes makes no guarantees at runtime that the underlying storage exists or is available.  High availability is left to the storage provider.
+
+## Goals
+
+* Allow administrators to describe available storage
+* Allow pod authors to discover and request persistence for their volumes
+* Enforce security through access control lists and securing storage to the same namespace as the pod volume
+* Enforce quotas through admission control
+* Ensure developers can rely on storage being available without being closely bound to a particular disk, server, network, or storage device.
+
+
+#### Allow admins to describe and create available storage
+
+Admins can describe storage using a `PersistentStoreController` for each type of VolumeSource being managed or create storage directly by using the `PersistentStore` API.  Storage created manually still requires a PersistentStoreController and accompanying interface to allow management by the cluster.
+
+
+```
+	
+	PersistentStore
+		TypeMeta
+		ObjectMeta
+			Name - internal name
+			Labels - labeled like Volume.PersistenceRequirements.Selector
+		Spec
+			VolumeSource -- will be backed by one of these:
+				HostPath (for a local disk implementation of storage)
+				GCEPersistentDisk { modes: ReadWriteOnce | ReadOnlyMany }
+				AWSElasticBlockStore { modes: ReadWriteOnce }
+				NFSMount { modes: ReadWriteOnce | ReadOnlyMany | ReadWriteMany }
+				*** many volume plugins
+			// volumes can be orphaned.  allow never delete.
+			RetentionPolicy StoreRetentionPolicy
+		Status
+			Phase
+			CurrentMounts	MountList
+			LastMounts		MountList
+			PodRef			ObjectReference -- the pod originally requesting persistence
+			
+
+	// Manages one type of PersistentStore
+	PersistentStoreController
+		Name
+		Spec
+			VolumeSource
+				GCEPersistentDisk { modes: ReadWriteOnce | ReadOnlyMany, resources: 5gb 30iops, .48 throughput  }
+			MaxInstances  100
+			IncrementBy   10
+		Status
+			InstanceCount
+			
+
+	//
+	//  addition to VolumeSource
+	//	
+	type MountMode byte
+	
+	const (
+		ReadWriteOnce MountMode = 1 << iota
+		ReadOnlyMany
+		ReadWriteMany
+	)
+	
+	func (vs *VolumeSource) canMount(mode MountMode) bool {
+		return vs.modes & mode != 0
+	}
+	
+```
+
+> To manually create storage, an admin would, for example, create a GCE VM, create, attach, and format 20 persistent disks.  The IDs are saved and the admin then creates 20 persistent-disk.json files and POSTs each GCE_PD to the API.  There are many ways to script adding stores to a cluster.
+
+### PersistentStoreController
+
+As a ReplicationController is to a Pod, so a PersistentStoreController is to one type of PersistentVolume.  A store controller manages all aspects of that volume's lifecycle.
```

Comment:
Yes, the analogy holds to repl controllers.  You would need 2 controllers, with each maintaining levels of that specific storage class.

But it would be preferable, I think, to have 1 per storage class and allow a list of template volumes.

This makes pool management easier, as the controller picks up all in that class (as manually made objects might not match a template volume).  It's also easier for the scheduler to ask 1 controller for advice based on type.

Which do you prefer?